README.md
[![All Contributors](https://img.shields.io/badge/all_contributors-1-orange.svg?style=flat-square)](#contributors)
[![Build Status](https://travis-ci.org/tagraha/address-book-react.svg?branch=master)](https://travis-ci.org/tagraha/address-book-react)
[![CircleCI](https://circleci.com/gh/tagraha/address-book-react.svg?style=svg)](https://circleci.com/gh/tagraha/address-book-react)
[![style: styled-components](https://img.shields.io/badge/style-%F0%9F%92%85%20styled--components-orange.svg?colorB=daa357&colorA=db748e)](https://github.com/styled-components/styled-components)

<a 
target="_blank"
href="https://david-dm.org/tagraha/address-book-react">
  <img src="https://david-dm.org/tagraha/address-book-react.svg" alt="Dependency Status" />
</a>
<a 
target="_blank"
href="https://david-dm.org/tagraha/address-book-react?type=dev">
  <img src="https://david-dm.org/tagraha/address-book-react/dev-status.svg" alt="Dev Dependency Status" />
</a>

## Demo
<a target="_blank" href="https://address-book-react.herokuapp.com/">https://address-book-react.herokuapp.com/</a>


### changelog
<a href="https://github.com/tagraha/address-book-react/blob/master/CHANGELOG.md">Changelog</a>

Q: are we using React Suspense or React Lazy feature?

A: No, it's a great feature though. Unfortunately, React Suspense is not support for Server Side Rendering at the moment
you can refer to this doc https://reactjs.org/docs/code-splitting.html

Heads up
--------
We are using react ^16.3.0. You might want to read this ([React v16.3.0: New lifecycles and context API](https://reactjs.org/blog/2018/03/29/react-v-16-3.html)) before starting playing around with this boilerplate.

Installation
------------
```bash
git clone git@github.com:tagraha/address-book-react.git
cd address-book-react
npm install
```

at this point make `.env` file on your root directory. you can copy it from `.env_example` file. after that you can continue run npm script below to start development environtment.

```bash
npm run develop
```

open up your browser and navigate to `localhost:1337`

Production script command
-------------------------
```bash
npm run build
npm start
```

Test script command
-------------------
```bash
npm run test
```

-------------

The DemoApp
-----------
Notice the `/shared` folder. that's where the DemoApp is located; inside `/container` folder. For the folder name, why is it called `shared`? because it will be used for compiling 2 bundles. `Server side` and `Client side`. `shared` folder is containing redux, components, container, and helper/utils. now let's take a look for `container` folder, shall we?

The main DemoApp is located in `/container/DemoApp/index.js`; we can define our initial metadata, css & js files there. and we can define all the website `route` there. it'll looks like something like this 
```javascript
<div className="main-app-container">
  <Switch>
    <Route exact path="/" component={PageHome} />
    <Route exact path="/setting" component={PageSetting} />
    <Route exact path="/about" component={PageAbout} />
    <Route component={PageError404} />
  </Switch>
</div>
```

Notice that the default website route (`/`) is rendering `PageHome` Component. here's the content of PageHome Component
```javascript
class PageHome extends PureComponent {
  render() {
    return (
      <Fragment>
        <Helmet>
          <title>Home</title>
        </Helmet>

        <div className="homepage-wrapper">
          <div className="fixed-wrapper">
            <div className="search-wrapper">
              <FilterInput />
            </div>
          </div>

          <UsersCard />

        </div>
      </Fragment>
    );
  }
}

export default PageHome
```

there are no state and props on  `PageHome` component because it's a container level file. the container shouldnt have any state or props in order to avoid unnecessary re-rendering/re-painting the whole page if there's any trivial state/props change.

Configuration
-------------
you can configure like disabling SSR mode, renaming service worker file etc with `value.js` file. here's the the quick look of the file

```javascript


const values = {
  clientConfigFilter: {
    // This is here as an example showing that you can expose variables
    // that were potentially provivded by the environment
    welcomeMessage: true,
    // We only need to expose the enabled flag of the service worker.
    serviceWorker: {
      enabled: true,
    },
    // We need to expose all the polyfill.io settings.
    polyfillIO: true,
    // We need to expose all the htmlPage settings.
    htmlPage: true,
  },

  // The host on which the server should run.
  host: EnvVars.string('HOST', 'localhost'),

  // The port on which the server should run.
  port: EnvVars.number('PORT', 1337),

  // The port on which the client bundle development server should run.
  clientDevServerPort: EnvVars.number('CLIENT_DEV_PORT', 7331),
  welcomeMessage: EnvVars.string('WELCOME_MSG', 'Hello world!'),

  // Disable server side rendering?
  disableSSR: false,
  browserCacheMaxAge: '365d',
}
```


CHANGELOG.md
#### 0.2.6 (2019-04-29)

##### Other Changes

* **Close Button:**  adding clsoe button on modal ([a971d8b5](https://github.com/tagraha/address-book-react/commit/a971d8b52375b97843508160e9877ebc894094d5))

#### 0.2.5 (2019-04-22)

##### Other Changes

* **readme:**  update readme ([67f1b291](https://github.com/tagraha/address-book-react/commit/67f1b2913fddc4adeab4baa6875ffafbda494227))

##### Tests

* **action:**  test onFilterChange actions ([f3f2f478](https://github.com/tagraha/address-book-react/commit/f3f2f478a67bc62110c656db8ab57bcb65cfe04d))
* **users:**  adding test for user loading ([971eff05](https://github.com/tagraha/address-book-react/commit/971eff05a93f0c26b73ba0e5f586f404c1aa2bb1))
* **jest:**  verbose tests ([ce724364](https://github.com/tagraha/address-book-react/commit/ce724364ce9e99cca6841db8777052f7a5b62682))

#### 0.2.4 (2019-04-22)

##### Other Changes

* **regex:**  make keyword filter incasesensitive ([436b95b4](https://github.com/tagraha/address-book-react/commit/436b95b489dc0f53004c2c668c7889cb8648e03e))

##### Tests

* **snapshot:**  update test snapshot for about page ([2435b928](https://github.com/tagraha/address-book-react/commit/2435b9285685aee4af7a62874234e5febe133499))

#### 0.2.3 (2019-04-21)

##### Bug Fixes

* **node:**  update node version ([3603e18e](https://github.com/tagraha/address-book-react/commit/3603e18e2b136b36c9654e23e9f0f04b06aceb1b))

##### Code Style Changes

* **modal:**  responsive modal ([57604e0e](https://github.com/tagraha/address-book-react/commit/57604e0e5dac28293a89fb639172a1a44e673ca9))

#### 0.2.2 (2019-04-21)

##### Bug Fixes

* **import:**  fix fail import caused by case sensitive ([977f62f7](https://github.com/tagraha/address-book-react/commit/977f62f71cdc4f3b00baaa01d0e6d50ef85e38c2))

#### 0.2.1 (2019-04-21)

##### New Features

* **limit:**  fetch limit to 1000 data ([a8b70fe5](https://github.com/tagraha/address-book-react/commit/a8b70fe50a15c68a247bb08e9d7e50970c7d37bd))

##### Other Changes

* **count:**  update load more user fetch count ([c10eed16](https://github.com/tagraha/address-book-react/commit/c10eed1606e0446ae95489147c3a4a7320c1691a))

##### Refactors

* **props:**  remove undefined props ([1dbe9430](https://github.com/tagraha/address-book-react/commit/1dbe9430f5cb227bce6b802b7e3517078221f6a9))
* **Modal:**  refactor home html structure ([8bbb0129](https://github.com/tagraha/address-book-react/commit/8bbb01294402dbcaf79479fd469949681f0a5f55))

##### Code Style Changes

* **image:**  make img clickable to open the modal ([9734bd7c](https://github.com/tagraha/address-book-react/commit/9734bd7c337055b93f24fd03bb553579ac417788))
* **Users Card:**  update users card style ([2f17b46a](https://github.com/tagraha/address-book-react/commit/2f17b46a198feae1f1ccb387f581d07d5005eae9))

##### Tests

* **snapshot:**  update snapshot ([1df038d6](https://github.com/tagraha/address-book-react/commit/1df038d68ca3a56a3cecc5d1384dca586e17a081))

### 0.2.0 (2019-04-21)

##### Refactors

* **perf:**  make PageHome not to re-render whole thing ([92593a2f](https://github.com/tagraha/address-book-react/commit/92593a2f2c29de4835c6a45f115d47db1e4c55be))

### 0.1.0 (2019-04-20)

##### New Features

* **filter:**  add basic filter functionality ([7c19faa7](https://github.com/tagraha/address-book-react/commit/7c19faa744376b90a66ad3bb45b12ae49e811346))
* **setting:**  add setting page ([750c082f](https://github.com/tagraha/address-book-react/commit/750c082f584f920e704bc8b9d7a8d2f1b7e13431))
* **users:**  basic feature of infinite scroll ([0322c438](https://github.com/tagraha/address-book-react/commit/0322c4381e537bf311d9f5f300d59aa4083ab2db))

##### Other Changes

* **security:**  whitelist randomuser.me domain for img ([9d6f7804](https://github.com/tagraha/address-book-react/commit/9d6f78044ee5e6e8b70be481563a8c7ad1cd9232))
* **package:**  add lodash.get ([90d420e3](https://github.com/tagraha/address-book-react/commit/90d420e3432b060c3ca9e492eb5588cdadc378e1))
* **icon:**  include fonts.googleapis.com to CSP whitelist ([a0b95693](https://github.com/tagraha/address-book-react/commit/a0b95693332190eba1d5dfaa437730b714ab7fe7))
* **file:**  adding gitignore ([c18588d9](https://github.com/tagraha/address-book-react/commit/c18588d98c98cd0a27ea42358550c70aafdbcb14))

##### Refactors

* **cleanup:**  cleanup the pages ([008ba55b](https://github.com/tagraha/address-book-react/commit/008ba55b99b22cfb9cd93967f860eb40a2f62e6c))

##### Tests

* **snapshot:**
  *  update test snapshot ([a39f87be](https://github.com/tagraha/address-book-react/commit/a39f87be58ed0b26776671819e3bb11bf95a43d1))
  *  update home route snapshot ([630b15cf](https://github.com/tagraha/address-book-react/commit/630b15cf5525efddaaf919c090d2ae37219ec4e2))
* **about:**  update test snapshot for about page ([b4eb061b](https://github.com/tagraha/address-book-react/commit/b4eb061b8f1e6f07dc8bf02e1ca9e40f846bab0f))



config/values.js
/**
 * Project Configuration.
 *
 * NOTE: All file/folder paths should be relative to the project root. The
 * absolute paths should be resolved during runtime by our build internal/server.
 */

import * as EnvVars from './utils/envVars';

const values = {
  // The configuration values that should be exposed to our client bundle.
  // This value gets passed through the /shared/utils/objects/filterWithRules
  // util to create a filter object that can be serialised and included
  // with our client bundle.
  clientConfigFilter: {
    // This is here as an example showing that you can expose variables
    // that were potentially provivded by the environment
    welcomeMessage: true,
    // We only need to expose the enabled flag of the service worker.
    serviceWorker: {
      enabled: true,
    },
    // We need to expose all the polyfill.io settings.
    polyfillIO: true,
    // We need to expose all the htmlPage settings.
    htmlPage: true,
  },

  // The host on which the server should run.
  host: EnvVars.string('HOST', 'localhost'),

  // The port on which the server should run.
  port: EnvVars.number('PORT', 1337),

  // The port on which the client bundle development server should run.
  clientDevServerPort: EnvVars.number('CLIENT_DEV_PORT', 7331),

  // This is an example environment variable which is used within the react
  // application to demonstrate the usage of environment variables across
  // the client and server bundles.
  welcomeMessage: EnvVars.string('WELCOME_MSG', 'Hello world!'),

  // Disable server side rendering?
  disableSSR: false,

  // How long should we set the browser cache for the served assets?
  // Don't worry, we add hashes to the files, so if they change the new files
  // will be served to browsers.
  // We are using the "ms" format to set the length.
  // @see https://www.npmjs.com/package/ms
  browserCacheMaxAge: '365d',

  // We use the polyfill.io service which provides the polyfills that a
  // client needs, which is far more optimal than the large output
  // generated by babel-polyfill.
  // Note: we have to keep this seperate from our "htmlPage" configuration
  // as the polyfill needs to be loaded BEFORE any of our other javascript
  // gets parsed.
  polyfillIO: {
    enabled: true,
    url: '//cdn.polyfill.io/v2/polyfill.min.js',
    // Reference https://qa.polyfill.io/v2/docs/features for a full list
    // of features.
    features: [
      // The default list.
      'default',
      'es6',
    ],
  },

  // Basic configuration for the HTML page that hosts our application.
  // We make use of react-helmet to consume the values below.
  // @see https://github.com/nfl/react-helmet
  htmlPage: {
    titleTemplate: 'React address book - %s',
    defaultTitle: 'React address book',
    description:
      'A starter kit giving you the minimum requirements for a production ready universal react application.',
  },

  // Content Security Policy (CSP)
  // @see server/middleware/security for more info.
  cspExtensions: {
    childSrc: [],
    connectSrc: [],
    defaultSrc: [],
    fontSrc: ['fonts.googleapis.com/css', 'fonts.gstatic.com'],
    imgSrc: ['https://randomuser.me'],
    mediaSrc: [],
    manifestSrc: [],
    objectSrc: [],
    scriptSrc: [
      // Allow scripts from cdn.polyfill.io so that we can import the
      // polyfill.
      'cdn.polyfill.io',
    ],
    styleSrc: [
      'cdn.rawgit.com/milligram/milligram/master/dist/milligram.min.css',
      'fonts.googleapis.com/css',
      'https://fonts.googleapis.com/',
    ],
  },

  // Path to the public assets that will be served off the root of the
  // HTTP server.
  publicAssetsPath: './public',

  // Where does our build output live?
  buildOutputPath: './build',

  // Do you want to included source maps for optimised builds of the client
  // bundle?
  includeSourceMapsForOptimisedClientBundle: false,

  // These extensions are tried when resolving src files for our bundles..
  bundleSrcTypes: ['js', 'jsx', 'json'],

  // What should we name the json output file that webpack generates
  // containing details of all output files for a bundle?
  bundleAssetsFileName: 'assets.json',

  // node_modules are not included in any bundles that target "node" as a
  // runtime (e.g.. the server bundle) as including them often breaks builds
  // due to thinks like require statements containing expressions..
  // However. some of the modules contain files need to be processed by
  // one of our Webpack loaders (e.g. CSS). Add any file types to the list
  // below to allow them to be processed by Webpack.
  nodeExternalsFileTypeWhitelist: [
    /\.(eot|woff|woff2|ttf|otf)$/,
    /\.(svg|png|jpg|jpeg|gif|ico)$/,
    /\.(mp4|mp3|ogg|swf|webp)$/,
    /\.(css|scss|sass|sss|less)$/,
  ],

  // Note: you can only have a single service worker instance.  Our service
  // worker implementation is bound to the "client" and "server" bundles.
  // It includes the "client" bundle assets, as well as the public folder assets,
  // and it is served by the "server" bundle.
  serviceWorker: {
    // Enabled?
    enabled: true,
    // Service worker name
    fileName: 'sw.js',
    // Paths to the public assets which should be included within our
    // service worker. Relative to our public folder path, and accepts glob
    // syntax.
    includePublicAssets: [
      // NOTE: This will include ALL of our public folder assets.  We do
      // a glob pull of them and then map them to /foo paths as all the
      // public folder assets get served off the root of our application.
      // You may or may not want to be including these assets.  Feel free
      // to remove this or instead include only a very specific set of
      // assets.
      './**/*',
    ],
    // Offline page file name.
    offlinePageFileName: 'offline.html',
  },

  bundles: {
    client: {
      // Src entry file.
      srcEntryFile: './client/index.js',

      // Src paths.
      srcPaths: [
        './client',
        './shared',
        // The service worker offline page generation needs access to the
        // config folder.  Don't worry we have guards within the config files
        // to ensure they never get included in a client bundle.
        './config',
      ],

      // Where does the client bundle output live?
      outputPath: './build/client',

      // What is the public http path at which we must serve the bundle from?
      webPath: '/client/',

      // Configuration settings for the development vendor DLL. This will be created
      // by our development server and provides an improved dev experience
      // by decreasing the number of modules that webpack needs to process
      // for every rebuild of our client bundle.  It by default uses the
      // dependencies configured in package.json however you can customise
      // which of these dependencies are excluded, whilst also being able to
      // specify the inclusion of additional modules below.
      devVendorDLL: {
        // Enabled?
        enabled: true,

        // Specify any dependencies that you would like to include in the
        // Vendor DLL.
        //
        // NOTE: It is also possible that some modules require specific
        // webpack loaders in order to be processed (e.g. CSS/SASS etc).
        // For these cases you don't want to include them in the Vendor DLL.
        include: [
          'react-async-component',
          'react',
          'react-dom',
          'react-helmet',
          'react-router-dom',
        ],

        // The name of the vendor DLL.
        name: '__dev_vendor_dll__',
      },
    },

    server: {
      // Src entry file.
      srcEntryFile: './server/index.js',

      // Src paths.
      srcPaths: ['./server', './shared', './config'],

      // Where does the server bundle output live?
      outputPath: './build/server',
    },
  },

  additionalNodeBundles: {
    // NOTE: The webpack configuration and build scripts have been built so
    // that you can add arbitrary additional node bundle configurations here.
    //
    // A common requirement for larger projects is to add additional "node"
    // target bundles (e.g an APi server endpoint). Therefore flexibility has been
    // baked into our webpack config factory to allow for this.
    //
    // Simply define additional configurations similar to below.  The development
    // server will manage starting them up for you.  The only requirement is that
    // within the entry for each bundle you create and return the "express"
    // listener.
    /*
    apiServer: {
      srcEntryFile: './api/index.js',
      srcPaths: [
        './api',
        './shared',
        './config',
      ],
      outputPath: './build/api',
    }
    */
  },

  // These plugin definitions provide you with advanced hooks into customising
  // the project without having to reach into the internals of the tools.
  //
  // We have decided to create this plugin approach so that you can come to
  // a centralised configuration folder to do most of your application
  // configuration adjustments.  Additionally it helps to make merging
  // from the origin starter kit a bit easier.
  plugins: {
    // This plugin allows you to provide final adjustments your babel
    // configurations for each bundle before they get processed.
    //
    // This function will be called once for each for your bundles.  It will be
    // provided the current webpack config, as well as the buildOptions which
    // detail which bundle and mode is being targetted for the current function run.
    babelConfig: (babelConfig, buildOptions) => {
      // eslint-disable-next-line no-unused-vars
      const { target, mode } = buildOptions;

      // Example
      /*
      if (target === 'server' && mode === 'development') {
        babelConfig.presets.push('foo');
      }
     */

      return babelConfig;
    },

    // This plugin allows you to provide final adjustments your webpack
    // configurations for each bundle before they get processed.
    //
    // I would recommend looking at the "webpack-merge" module to help you with
    // merging modifications to each config.
    //
    // This function will be called once for each for your bundles.  It will be
    // provided the current webpack config, as well as the buildOptions which
    // detail which bundle and mode is being targetted for the current function run.
    webpackConfig: (webpackConfig, buildOptions) => {
      // eslint-disable-next-line no-unused-vars
      const { target, mode } = buildOptions;

      // Example:
      /*
      if (target === 'server' && mode === 'development') {
        webpackConfig.plugins.push(new MyCoolWebpackPlugin());
      }
      */

      // Debugging/Logging Example:
      /*
      if (target === 'server') {
        console.log(JSON.stringify(webpackConfig, null, 4));
      }
      */

      return webpackConfig;
    },
  },
};

// This protects us from accidentally including this configuration in our
// client bundle. That would be a big NO NO to do. :)
if (process.env.BUILD_FLAG_IS_CLIENT === 'true') {
  throw new Error(
    "You shouldn't be importing the `<projectroot>/config/values.js` directly into code that will be included in your 'client' bundle as the configuration object will be sent to user's browsers. This could be a security risk! Instead, use the `config` helper function located at `<projectroot>/config/index.js`.",
  );
}

export default values;
